### PR TITLE
Important updates to Balancer Trades to improve `dex.trades`

### PR DESCRIPTION
**Pull request by thetroyharris**
Brief comments on the purpose of your changes:

Updating the Balancer V2 Trades Spell logic to accurately price trades involving Balancer Pool Tokens. I recently updated the Balancer Pool Tokens Spell, #3416, and am now updating the trades Spell to take advantage of the new pricing table.

Additionally, I did some Spell aggregation cleanup regarding the Balancer Trades and Balancer V2 Trades Spells.

cc: @mendesfabio @markusbkoch 

**For Dune Engine V2**

I've checked that:

### General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [x] `coin_id` represents the ID of the coin on coinpaprika.com
* [x] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [x] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [x] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [x] where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by thetroyharris** [comment]
This unique combination of columns error has me quite confused. Any help would be greatly appreciated. It use to be the same combination as the other chains that passed the test, but I started changing it up to see if it would pass... no luck.

I even took the results from the test and cross checked a few of the tx_hashes against the working FE query and some of them only return 1 result. Which does not make any sense to me.

Thanks in advance!

**Comment by mendesfabio** [comment]
> This unique combination of columns error has me quite confused. Any help would be greatly appreciated. It use to be the same combination as the other chains that passed the test, but I started changing it up to see if it would pass... no luck.
> 
> I even took the results from the test and cross checked a few of the tx_hashes against the working FE query and some of them only return 1 result. Which does not make any sense to me.
> 
> Thanks in advance!

hey were you able to query the test failures? I can't run it on Dune 🤔

**Comment by thetroyharris** [comment]
> > This unique combination of columns error has me quite confused. Any help would be greatly appreciated. It use to be the same combination as the other chains that passed the test, but I started changing it up to see if it would pass... no luck.
> > I even took the results from the test and cross checked a few of the tx_hashes against the working FE query and some of them only return 1 result. Which does not make any sense to me.
> > Thanks in advance!
> 
> hey were you able to query the test failures? I can't run it on Dune 🤔

I was able to yesterday, however I can't run the query today either. I changed `block_date` to `block_time` and committed the changes so we could get a fresh test table.

**Comment by mendesfabio** [comment]
> > > This unique combination of columns error has me quite confused. Any help would be greatly appreciated. It use to be the same combination as the other chains that passed the test, but I started changing it up to see if it would pass... no luck.
> > > I even took the results from the test and cross checked a few of the tx_hashes against the working FE query and some of them only return 1 result. Which does not make any sense to me.
> > > Thanks in advance!
> > 
> > 
> > hey were you able to query the test failures? I can't run it on Dune 🤔
> 
> I was able to yesterday, however I can't run the query today either. I changed `block_date` to `block_time` and committed the changes so we could get a fresh test table.
> 
> EDIT: test results are working now `select * from test_schema.git_150adc26_dbt_utils_unique_combination_of_columns_balancer_v2_ethereum_trades_blockchain__block_time__tx_hash__evt_index`

I investigated one case and found this duplicated swap transaction with 2 diff swap fees. It seems that the pool had 2 fees set in the same block and since we're left joining with `swap_fees` it gets duplicated - guess the other cases may have the same root cause. We have to handle it since it will happen frequently with this [CoWSwap integration](https://twitter.com/CoWSwap/status/1668973199340748800)

image

image

**Comment by thetroyharris** [comment]
> > > > This unique combination of columns error has me quite confused. Any help would be greatly appreciated. It use to be the same combination as the other chains that passed the test, but I started changing it up to see if it would pass... no luck.
> > > > I even took the results from the test and cross checked a few of the tx_hashes against the working FE query and some of them only return 1 result. Which does not make any sense to me.
> > > > Thanks in advance!
> > > 
> > > 
> > > hey were you able to query the test failures? I can't run it on Dune 🤔
> > 
> > 
> > I was able to yesterday, however I can't run the query today either. I changed `block_date` to `block_time` and committed the changes so we could get a fresh test table.
> > EDIT: test results are working now `select * from test_schema.git_150adc26_dbt_utils_unique_combination_of_columns_balancer_v2_ethereum_trades_blockchain__block_time__tx_hash__evt_index`
> 
> I investigated one case and found this duplicated swap transaction with 2 diff swap fees. It seems that the pool had 2 fees set in the same block and since we're left joining with `swap_fees` the swap gets duplicated - guess the other cases may have the same root cause. We have to handle it since it will happen frequently with this [CoWSwap integration](https://twitter.com/CoWSwap/status/1668973199340748800)
> 
> image image

Digging in now, thank you!

**Comment by thetroyharris** [comment]
The previous issue we had, duplicate entires, occurred when a pool's `swap_fee_percentage` was changed multiple times in the same txn or was changed in the same txn as the swap.

The solution was to two fold:
1.  Get a table of swap fees where, if a swap fee changed multiple times in a txn, we grab the highest evt number change in that txn. This solved the issue when a swap fee is changed multiple times in a txn AND that txn does not contain a swap event.
2. Other scenario is a swap fee is reduced prior to a trade being executed. In this instance the swap fee change and token swap occur in the same txn so we can grab the swap fee change event before the swap and use that as the proper swap fee.

**Comment by thetroyharris** [comment]
@jeff-dude 

This PR is ready for review.

**Comment by thetroyharris** [comment]
@jeff-dude 

All checks have passed. :)

File: models/balancer/arbitrum/balancer_v2_arbitrum_trades.sql [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+                                "project",
+                                "balancer_v2",
+                                \'["mendesfabio", "jacektrocinski", "thetroyharris"]\') }}'
+    )
+}}
+
+{% set project_start_date = '2021-08-27' %}
+
+WITH    
+    fees_base AS (SELECT *, MAX(index) OVER(PARTITION BY tx_hash, contract_address) AS max_index_same_tx FROM balancer_v2_arbitrum.pools_fees),
```
Comment (jeff-dude) [inline models/balancer/arbitrum/balancer_v2_arbitrum_trades.sql:19]: please use jinja syntax on all sources / references throughout. another example is line 27
@thetroyharris

File: models/balancer/arbitrum/balancer_v2_arbitrum_trades.sql [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                AND most.block_time = swap_fees.max_fee_evt_block_time
+                AND most.tx_hash <> swap_fees.evt_tx_hash
+            LEFT JOIN edge_case_fees edge
+                ON edge.evt_block_number = swap_fees.evt_block_number
+                AND edge.evt_tx_hash = swap_fees.evt_tx_hash
+                AND edge.evt_index = swap_fees.evt_index
+        WHERE
+            swap.tokenIn <> swap_fees.contract_address
+            AND swap.tokenOut <> swap_fees.contract_address
+            {% if is_incremental() %}
```
Comment (jeff-dude) [inline models/balancer/arbitrum/balancer_v2_arbitrum_trades.sql:80]: can the incremental filters be put in the CTE called?

**Review by jeff-dude — CHANGES_REQUESTED** [review]
few comments below. i only looked at arbitrum, but likely applies universally

File: models/balancer/arbitrum/balancer_v2_arbitrum_trades.sql (hunk 1 above)
Comment (thetroyharris) [reply models/balancer/arbitrum/balancer_v2_arbitrum_trades.sql:19]: On it. Apologies for that as it was a late addition. Will make sure to better check for the jinja syntax moving forward.

File: models/balancer/arbitrum/balancer_v2_arbitrum_trades.sql (hunk 2 above)
Comment (thetroyharris) [reply models/balancer/arbitrum/balancer_v2_arbitrum_trades.sql]: Yes, done.

File: models/balancer/arbitrum/balancer_v2_arbitrum_trades.sql (hunk 1 above)
Comment (thetroyharris) [reply models/balancer/arbitrum/balancer_v2_arbitrum_trades.sql]: Done across all chains, thank you.

**Review by jeff-dude — APPROVED** [review]
thank you!

**Pull request merged**